File: services/api/src/bills/capture-bill-split.ts
import { ApiError } from "../http/errors";
import type { BillShareInput } from "./bill-split";
import type { ParsedReceipt } from "../receipts/receipt.types";

export type CaptureBillItemInput = {
  name: string;
  quantity: number;
  unitPriceCents: number;
  totalPriceCents: number;
  assignedUserIds: string[];
};

function dollarsToCents(value: number | null | undefined): number {
  if (value == null) {
    return 0;
  }
  return Math.round(value * 100);
}

function allocateRemainderCents(
  orderedUserIds: string[],
  baseAmounts: Map<string, number>,
  remainderCents: number,
): Map<string, number> {
  const result = new Map(baseAmounts);
  for (let i = 0; i < remainderCents; i++) {
    const userId = orderedUserIds[i % orderedUserIds.length];
    result.set(userId, (result.get(userId) ?? 0) + 1);
  }
  return result;
}

export function computeCaptureShares(
  receipt: ParsedReceipt,
  items: CaptureBillItemInput[],
  participantIds: string[],
): { shares: BillShareInput[]; totalCents: number } {
  const participants = [...new Set(participantIds)].sort();
  if (participants.length === 0) {
    throw new ApiError(400, "INVALID_PARTICIPANTS", "At least one participant is required");
  }

  const baseCents = new Map<string, number>();
  for (const userId of participants) {
    baseCents.set(userId, 0);
  }

  let itemsSubtotalCents = 0;

  for (const item of items) {
    if (item.assignedUserIds.length === 0) {
      throw new ApiError(400, "UNASSIGNED_ITEM", `Item "${item.name}" has no assignees`);
    }

    const assignees = [...new Set(item.assignedUserIds)].sort();
    for (const assignee of assignees) {
      if (!participants.includes(assignee)) {
        throw new ApiError(400, "INVALID_ASSIGNEE", "Assignee must be a participant");
      }
    }

    itemsSubtotalCents += item.totalPriceCents;
    const perAssignee = Math.floor(item.totalPriceCents / assignees.length);
    const remainder = item.totalPriceCents % assignees.length;

    assignees.forEach((userId, index) => {
      const extra = index < remainder ? 1 : 0;
      baseCents.set(userId, (baseCents.get(userId) ?? 0) + perAssignee + extra);
    });
  }

  const taxCents = dollarsToCents(receipt.tax);
  const tipCents = dollarsToCents(receipt.tip);
  const extrasCents = taxCents + tipCents;

  const extrasBase = new Map<string, number>();
  for (const userId of participants) {
    extrasBase.set(userId, 0);
  }

  if (extrasCents > 0) {
    if (itemsSubtotalCents > 0) {
      let allocated = 0;
      const ordered = [...participants].sort();
      for (let i = 0; i < ordered.length; i++) {
        const userId = ordered[i];
        const userBase = baseCents.get(userId) ?? 0;
        if (i === ordered.length - 1) {
          extrasBase.set(userId, extrasCents - allocated);
        } else {
          const share = Math.floor((extrasCents * userBase) / itemsSubtotalCents);
          extrasBase.set(userId, share);
          allocated += share;
        }
      }
    } else {
      const perUser = Math.floor(extrasCents / participants.length);
      const remainder = extrasCents % participants.length;
      const allocated = allocateRemainderCents(
        participants,
        new Map(participants.map((id) => [id, perUser])),
        remainder,
      );
      for (const [userId, amount] of allocated) {
        extrasBase.set(userId, amount);
      }
    }
  }

  const totalCents =
    receipt.total != null ? dollarsToCents(receipt.total) : itemsSubtotalCents + extrasCents;

  const shares: BillShareInput[] = participants.map((userId) => ({
    userId,
    shareCents: (baseCents.get(userId) ?? 0) + (extrasBase.get(userId) ?? 0),
  }));

  const shareSum = shares.reduce((sum, share) => sum + share.shareCents, 0);
  if (shareSum !== totalCents) {
    const diff = totalCents - shareSum;
    const payerCandidate = shares.find((s) => s.shareCents > 0) ?? shares[0];
    payerCandidate.shareCents += diff;
  }

  return { shares, totalCents };
}

export function parseReceiptIncurredAt(receipt: ParsedReceipt): Date {
  if (receipt.date) {
    const timePart = receipt.time ? ` ${receipt.time}` : "";
    const parsed = Date.parse(`${receipt.date}${timePart}`);
    if (!Number.isNaN(parsed)) {
      return new Date(parsed);
    }
  }
  return new Date();
}

File: apps/Mobile-App/split_the_bill/lib/utils/capture_bill_split.dart
import '../models/receipt.dart';

class CaptureShareDraft {
  const CaptureShareDraft({required this.userId, required this.shareCents});

  final String userId;
  final int shareCents;
}

class CaptureShareResult {
  const CaptureShareResult({
    required this.shares,
    required this.totalCents,
  });

  final List<CaptureShareDraft> shares;
  final int totalCents;
}

int _dollarsToCents(double? value) {
  if (value == null) return 0;
  return (value * 100).round();
}

CaptureShareResult computeCaptureShares({
  required ParsedReceipt receipt,
  required List<ReceiptItem> items,
  required Map<int, Set<String>> assignments,
  required List<String> participantIds,
}) {
  final participants = participantIds.toSet().toList()..sort();
  final baseCents = <String, int>{for (final id in participants) id: 0};

  var itemsSubtotalCents = 0;

  for (var index = 0; index < items.length; index++) {
    final item = items[index];
    final assignees = (assignments[index] ?? {}).toList()..sort();
    if (assignees.isEmpty) {
      throw StateError('Item "${item.name}" has no assignees');
    }

    itemsSubtotalCents += item.totalPriceCents;
    final perAssignee = item.totalPriceCents ~/ assignees.length;
    final remainder = item.totalPriceCents % assignees.length;

    for (var i = 0; i < assignees.length; i++) {
      final userId = assignees[i];
      final extra = i < remainder ? 1 : 0;
      baseCents[userId] = (baseCents[userId] ?? 0) + perAssignee + extra;
    }
  }

  final taxCents = _dollarsToCents(receipt.tax);
  final tipCents = _dollarsToCents(receipt.tip);
  final otherFeesCents = _dollarsToCents(receipt.otherFees);
  final extrasCents = taxCents + tipCents + otherFeesCents;
  final extrasBase = <String, int>{for (final id in participants) id: 0};

  if (extrasCents > 0) {
    if (itemsSubtotalCents > 0) {
      var allocated = 0;
      for (var i = 0; i < participants.length; i++) {
        final userId = participants[i];
        final userBase = baseCents[userId] ?? 0;
        if (i == participants.length - 1) {
          extrasBase[userId] = extrasCents - allocated;
        } else {
          final share = ((extrasCents * userBase) / itemsSubtotalCents).floor();
          extrasBase[userId] = share;
          allocated += share;
        }
      }
    } else {
      final perUser = extrasCents ~/ participants.length;
      final remainder = extrasCents % participants.length;
      for (var i = 0; i < participants.length; i++) {
        extrasBase[participants[i]] = perUser + (i < remainder ? 1 : 0);
      }
    }
  }

  final totalCents = receipt.total != null
      ? _dollarsToCents(receipt.total)
      : itemsSubtotalCents + extrasCents;

  final shares = participants
      .map(
        (userId) => CaptureShareDraft(
          userId: userId,
          shareCents: (baseCents[userId] ?? 0) + (extrasBase[userId] ?? 0),
        ),
      )
      .toList();

  final shareSum = shares.fold<int>(0, (sum, share) => sum + share.shareCents);
  if (shareSum != totalCents && shares.isNotEmpty) {
    final diff = totalCents - shareSum;
    final first = shares.first;
    shares[0] = CaptureShareDraft(
      userId: first.userId,
      shareCents: first.shareCents + diff,
    );
  }

  return CaptureShareResult(shares: shares, totalCents: totalCents);
}

List<({ReceiptItem item, int shareCents})> itemsForUser({
  required List<ReceiptItem> items,
  required Map<int, Set<String>> assignments,
  required String targetUserId,
}) {
  final result = <({ReceiptItem item, int shareCents})>[];
  for (var index = 0; index < items.length; index++) {
    final assignees = assignments[index] ?? {};
    if (!assignees.contains(targetUserId)) continue;

    final item = items[index];
    final perAssignee = item.totalPriceCents ~/ assignees.length;
    final remainder = item.totalPriceCents % assignees.length;
    final sorted = assignees.toList()..sort();
    final position = sorted.indexOf(targetUserId);
    final extra = position < remainder ? 1 : 0;
    result.add((item: item, shareCents: perAssignee + extra));
  }
  return result;
}

int userItemSubtotalCents({
  required List<ReceiptItem> items,
  required Map<int, Set<String>> assignments,
  required String targetUserId,
}) {
  return itemsForUser(
    items: items,
    assignments: assignments,
    targetUserId: targetUserId,
  ).fold<int>(0, (sum, entry) => sum + entry.shareCents);
}
